docs/dx.rst

```diff
@@ -0,0 +1,187 @@
+.. _dx:
+
+DNAnexus
+========
+
+Canonical Paths on DNAnexus
+---------------------------
+Every file resource on DNAnexus is given a canonical ID to refer it by. This is
+internally given by the platform and is not determined by stor. However, stor
+does provide the option of referring to the resource through their canonical IDs.
+Each individual project and file has a canonical ID associated with it. This
+ID consists of 24 case-sensitive characters from ``0123456789BFGJKPQVXYZbfgjkpqvxyz``
+prepended with 'project-' or 'file-' respectively. Example canonical IDs are
+``project-j47b1k3z8Jqqv001213v312j1``, and ``file-47jK67093475061g3v95369p``.
+
+Each individual file with its unique ID, can manifest in only one project on
+DNAnexus. This is because the metadata associated with the file in that project
+can only have one path to the file. Also, since the canonicalized path to the
+file on DNAnexus is represented by::
+
+    'project-j47b1k3z8Jqqv001213v312j1:file-47jK67093475061g3v95369p'
+
+each project can only have each canonical file at one location within the project.
+However, one canonical file can be present in multiple projects at different paths.
+
+Thus, stor has two subclass implementations of `DXPath`: `DXCanonicalPath` and
+`DXVirtualPath`. As the name suggests, `DXCanonicalPath` deals with paths like::
+
+    Path('dx://project-j47b1k3z8Jqqv001213v312j1:file-47jK67093475061g3v95369p')
+    OR
+    Path('dx://project-j47b1k3z8Jqqv001213v312j1')
+
+`DXVirtualPath` handles paths that have any human readable element in them::
+
+    Path('dx://project-j47b1k3z8Jqqv001213v312j1:/path/to/file.txt')
+    OR
+    Path('dx://myproject:/path/to/file.txt')
+
+You can obtain the `DXCanonicalPath` for a resource from a `DXVirtualPath`
+and vice versa.
+
+Projects on DNAnexus
+--------------------
+All resources on the DX platform are organized into projects, which is the main
+container of the resource. The access and billing etc permissions to the resource
+are determined by (and the same as) the user's permission to the project.
+
+On stor, the project is the first part of any `DXPath` and is unique because
+it must end in a trailing ':'. Every `DXPath` must have a project associated.
+More on this later.
+
+Directories on DNAnexus
+-----------------------
+DNAnexus has the concept of directories on the platform like posix (unlike
```

Comment:
this should be made into a table about POSIX vs. Swift vs. S3 vs. DNAnexus

Reply:
diff between POSIX and DNAnexus:

1. multiple files can have same virtual path
2. folder and file can be at same name

Reply:
(and (1) behavior should be documented too)